proto/src/main.cpp

```diff
@@ -1,73 +1,89 @@
 
+SYSTEM_MODE(AUTOMATIC);
 SYSTEM_THREAD(ENABLED);
 
-void onSerialData();
-JsonMaker json_maker;
-Sensor_RSSI rssi;
-Sensor_ECU ecu(&Serial1);
-
-
-uint32_t last_blink = 0; // Time of last blink
-uint32_t last_publish = 0; // Time of last publish
-boolean led_state = LOW;
+// Disable publish when not necessary during development to reduce data usage
+#define PUBLISH_ENABLED             1
+#define PUBLISH_INTERVAL_MS         5000
+
+#define GPS_UPDATE_INTERVAL_MS      1000
+#define THERMO_UPDATE_INTERVAL_MS   500
+
+JsonMaker jsonMaker;
+SensorEcu ecu(&Serial1);
+SensorGps gps(GPS_UPDATE_INTERVAL_MS);
+SensorThermo thermoA(&SPI, A5, THERMO_UPDATE_INTERVAL_MS);
+
+uint32_t lastPublish = 0;
+
+/**
+ * Publishes a new message to Particle Cloud
+ * */
+void publishMessage(){
+    // Create JSON object for publish
+    jsonMaker.refresh();
+    // ECU data
+    jsonMaker.add("PROTO-ECT", ecu.getECT());
+    jsonMaker.add("PROTO-IAT", ecu.getIAT());
+    jsonMaker.add("PROTO-RPM", ecu.getRPM());
+    jsonMaker.add("PROTO-UBADC", ecu.getUbAdc());
+    jsonMaker.add("PROTO-O2S", ecu.getO2S());
+    jsonMaker.add("PROTO-SPARK", ecu.getSpark());
+    // GPS data
+    jsonMaker.add("PROTO-Location", gps.getSentence());
+    jsonMaker.add("PROTO-Speed", gps.getSpeedKph());
+
+    if(PUBLISH_ENABLED){
+        // Publish to Particle Cloud
+        Particle.publish("Proto", jsonMaker.get(), PRIVATE, WITH_ACK);
+        Serial.println("Publish - ENABLED - Message: ");
+    }else{
+        Serial.println("Publish - DISABLED - Message: ");
+    }
 
-Timer timer(1, onSerialData);
+    Serial.println("New JSON Message: " + jsonMaker.get());
 
-// Handler for any new message on ID "Proto"
-void proto_response(const char *event, const char *data) {
-    Serial.println("Received Message on ID: " + String(event) + " - Message Data: " + String(data));
+    // Any sensors that are working but not yet packaged for publish
+    Serial.println("Not in Message: ");
+    Serial.print("Current Temperature (Thermo1): "); Serial.print(thermoA.getTemp()); Serial.println("C");
+    Serial.println();
 }
 
+/**
+ * 
+ * SETUP
+ * 
+ * */
 void setup() {
-    pinMode(D7, OUTPUT);
+    Serial.begin(115200);
 
     ecu.begin();
-    timer.start();
-
-    // Subscribe to any new messages on ID "Proto"
-    Particle.subscribe("hook-response/Proto", proto_response, MY_DEVICES);
+    gps.begin();
+    thermoA.begin();
 
-    // Setup function only runs after Boron connected in (default) Automatic mode
-    Serial.println("Particle Connected!");
 }
 
+/**
+ * 
+ * LOOP
+ * 
+ * */
 void loop() {
-
-    digitalWrite(D7, led_state);
-    
-    // Blink the LED
-    if (led_state & (millis() - last_blink >= BLINK_INTERVAL_ON)){
-        led_state = LOW;
-        last_blink = millis();
-    }else if(!led_state & (millis() - last_blink >= BLINK_INTERVAL_OFF)){
-        led_state = HIGH;
-        last_blink = millis();
+    ecu.handle();
+    gps.handle();
+    thermoA.handle();
```

Comment:
Would you be able to time (just diff the `millis()` before and after each call) each of these handlers? Ideally they should be quite fast (so it's not a problem), but if there's a tangible delay between each then our single datapoint for all the data from this timestamp would not line up.

Reply:
I added optional timing that you can enable with a flag in the latest commit. Here are the values we're getting (in MICROseconds, not miliseconds): 

TIME ELAPSED FOR: 
ECU Polling:                   537us
GPS Polling:                11158us
Thermo Polling:             673us
Build JSON Message: 1559us